---
title: "GC normalization by CANOES."
output: html_notebook
Author: Renjie Tan
---

############################################################################
# STEP 1: Fetch the directory locations and file names from the shell script
############################################################################
```{r}
#library(doParallel)
#library(foreach)
biocluster_home_path  <- '/Users/rtan/biocluster_home'
biocluster_terra_path <- '/Users/rtan/terra'
scripts_dir <- paste(biocluster_home_path, '/cnv_toolkit/scripts/rscripts/'  ,sep='')
data_dir    <- paste(biocluster_terra_path,'/SPARK/0-data/RC/spark_merged_RC/',sep='')
batch_name  <- 'spark1'
output_dir  <- '/Users/rtan/biocluster_home/CNV_Espresso/'
cores       <- '1'
source(paste(scripts_dir,"CANOES.R",sep=""))

# sample list
sample_name_file <- paste(data_dir,'merged_sample_list_', batch_name,'.list',sep='')

#output file
output_dir <- paste(output_dir, batch_name,'/',sep='')
output_file_name <- paste(output_dir,'canoes_calls_',batch_name,'.csv',sep='')

#log file
log_dir <- paste(output_dir,'/log/',sep='')
if(!dir.exists(output_dir)){
	dir.create(output_dir)
	dir.create(log_dir)
}
```

######################################################
# STEP 2: Read the read count file and GC content file
######################################################
```{r}
# gc file
gc_file <- paste(data_dir, 'gc.txt.bz2',sep='')
file.exists(gc_file)
gc <- read.table(bzfile(gc_file))$V2

# input file
input_file_name <- paste(data_dir, 'cons_canoes_reads_',batch_name,'.bz2',sep='') 
file.exists(input_file_name)
canoes.reads <- read.table(bzfile(input_file_name))
dim(canoes.reads)
```
########################################################
# STEP 3: Assign column names and merge columns together
########################################################
```{r}
num_cols <- dim(canoes.reads)[2]
num_samples <- num_cols - 3
sample.names <- read.table(file=sample_name_file)
sample.names <- sub(".tar.bz2","",sample.names[,1])
sample.names <- sub(".cram","",sample.names)
sample.names <- sub(".bam","",sample.names)
names(canoes.reads) <- c("chromosome", "start", "end", sample.names)
target <- seq(1, nrow(canoes.reads))
canoes.reads <- cbind(target, canoes.reads[,1:3], gc, canoes.reads[,-c(1,2,3)])
```
```{r}
test_sample.name="SP0000203"
which(colnames(canoes.reads)==test_sample.name)
canoes.reads[1:10,c(1,2,3,4,5,6,25)]
```

######################################################
# [TBD] Here need to draw the GC plots for raw GC
######################################################
```{r}
RC_beforeNorm <- canoes.reads
canoes.reads_test <- canoes.reads[1:8000,]
dim(canoes.reads_test)
plot(canoes.reads_test$gc,canoes.reads_test$SP0000203, ylim = c(0,500))
```
```{r}
range(RC_beforeNorm$SP0000203)
loessMod50 <- loess(SP0000203 ~ gc, data=RC_beforeNorm, span=0.50) # 50% smoothing span
# get smoothed output
smoothed50 <- predict(loessMod50) 
# Plot it
i <- ggplot(RC_beforeNorm, aes(gc, smoothed50))
i + geom_line()
```

```{r}
# RT moved the following scripts outside of CallCNVs function in CANOES.R.
######################################################################################################
if(class(canoes.reads$chromosome) != "numeric"){
    canoes.reads$chromosome <- as.character(canoes.reads$chromosome) #RT modifided
}

if (length(setdiff(names(canoes.reads)[1:5], c("target", "chromosome", "start", "end", "gc"))) > 0){
	stop("First five columns of counts matrix must be target, chromosome, start, end, gc")
}
if (length(setdiff(unique(canoes.reads$chromosome), seq(1:22))) > 0) {
	# remove sex chromosomes
	cat("Trying to remove sex chromosomes and 'chr' prefixes\n")
	canoes.reads <- subset(canoes.reads, !chromosome %in% c("chrX", "chrY", "X", "Y"))
	if (sum(grepl("chr", canoes.reads$chromosome))==length(canoes.reads$chromosome)){
	  canoes.reads$chromosome <- gsub("chr", "", canoes.reads$chromosome)
	}
	canoes.reads$chromosome <- as.numeric(canoes.reads$chromosome)
	if (length(setdiff(unique(canoes.reads$chromosome), seq(1:22))) > 0) 
	  stop("chromosome must take value in range 1-22 (support for sex chromosomes to come)")
}

if(class(canoes.reads$chromosome) != "numeric"){
    canoes.reads$chromosome <- as.numeric(as.character(canoes.reads$chromosome)) #RT modifided
}
dim(canoes.reads)
######################################################################################################
```

```{r}
# calculate covariance of read count across samples
xcnv.list <- vector('list', length(sample.names))
file_name <- strsplit(basename(input_file_name),"\\.")[[1]][1]
cov_file  <- paste(data_dir,'cov.',file_name,sep='')
if (file.exists(cov_file)){
	load(file=cov_file)
}
if (!exists('covariance')){
  cat("Please calculate covariance of read count across samples at first.")
#	cat(paste('[',format(Sys.time()),']'," calculate covariance of read count across samples\n",sep=''))#
#	system.time(covariance <- cor(canoes.reads[, sample.names], canoes.reads[, sample.names]))
#	save(covariance,file=paste('./cov',basename(input_file_name),sep='.'))
}
```

######################################################
# STEP 4: Load R script and process each column/sample
######################################################
```{r}
sample.name="SP0000203"
log_file <- paste(log_dir,'log_core.log',sep='')
#xcnv.list[[x]] <- CallCNVs(sample.name, canoes.reads, covariance,log_file)

p=1e-08
Tnum=6
D=70000
numrefs=30
get.dfs=F
homdel.mean=0.2
counts <- canoes.reads
if (!sample.name %in% names(counts)){stop("No column for sample ", sample.name, " in counts matrix")}
if (length(setdiff(names(counts)[1:5], c("target", "chromosome", "start", "end", "gc"))) > 0){
  stop("First five columns of counts matrix must be target, chromosome, start, end, gc")
}

#######################################################################################################
if(class(counts$chromosome) != "numeric"){
  counts$chromosome <- as.numeric(as.character(counts$chromosome)) #RT modifided
}
#######################################################################################################

library(plyr)
counts <- arrange(counts, chromosome, start)
dim(counts)
counts[1:10,c(1,2,3,4,5,6,25)]
```
```{r}
if (p <= 0){
  stop("parameter p must be positive")
}
if (Tnum <= 0){
  stop("parameter Tnum must be positive")
}
if (D <= 0){
  stop("parameter D must be positive")
}
if (numrefs <= 0){
  stop("parameter numrefs must be positive")
}
sample.names <- colnames(counts)[-seq(1,5)]
length(sample.names)
```

Select 500 samples for the test to avoid memory corruption
```{r}
sample.names <- sample.names[1:500]
length(sample.names)
```

```{r}
# find mean coverage of probes
write.table(paste('[',format(Sys.time()),'] [',sample.name,'] find mean coverage of probes',sep=''), file=log_file, col.names=FALSE, row.names = FALSE, append=TRUE)
cat(paste('[',format(Sys.time()),']'," find mean coverage of probes\n",sep=''))
mean.counts <- mean(apply(counts[, sample.names], 2, mean))
mean.counts
```
```{r}
# normalize counts; round so we can use negative binomial
write.table(paste('[',format(Sys.time()),'] [',sample.name,'] normalize counts; round so we can use negative binomial',sep=''), file=log_file, col.names=FALSE, append=TRUE)
cat(paste('[',format(Sys.time()),']'," normalize counts; round so we can use negative binomial\n",sep=''))
gc()
counts[, sample.names] <- apply(counts[, sample.names], 2, 
      function(x, mean.counts) 
               round(x * mean.counts / mean(x)), mean.counts)

# calculate covariance of read count across samples
### Removed to outside of CallCNVs function
reference.samples <- setdiff(sample.names, sample.name)
covariances <- covariance[sample.name, reference.samples]
reference.samples <- names(sort(covariances, 
        decreasing=T)[1:min(numrefs, length(covariances))])
reference.samples
```
```{r}
write.table(paste('[',format(Sys.time()),'] [',sample.name,'] normalize counts; round so we can use negative binomial',sep=''), file=log_file, col.names=FALSE, append=TRUE)
cat(paste('[',format(Sys.time()),']'," Calculating the mean RC ...",sep=''))
sample.mean.counts <- mean(counts[, sample.name])
sample.sumcounts <- apply(counts[, reference.samples], 2, sum)
cat(sample.mean.counts)
cat(sample.sumcounts)
```
```{r}
# normalize reference samples to sample of interest
write.table(paste('[',format(Sys.time()),'] [',sample.name,'] normalize counts; round so we can use negative binomial',sep=''), file=log_file, col.names=FALSE, append=TRUE)
cat(paste('[',format(Sys.time()),']'," normalize reference samples to sample of interest\n",sep=''))
counts[, reference.samples] <- apply(counts[, reference.samples], 2, 
      function(x, sample.mean.counts) 
              round(x * sample.mean.counts / 
              mean(x)), sample.mean.counts)  
```
```{r}
# select reference samples and weightings using non-negative least squares
write.table(paste('[',format(Sys.time()),'] [',sample.name,'] select reference samples and weightings using non-negative least squares',sep=''), file=log_file, col.names=FALSE, append=TRUE)
cat(paste('[',format(Sys.time()),']'," select reference samples and weightings using non-negative least squares\n",sep=''))
b <- counts[, sample.name]
A <- as.matrix(counts[, reference.samples])
library(nnls)
all <- nnls(A, b)$x
est <- matrix(0, nrow=50, ncol=length(reference.samples))
set.seed(1)
for (i in 1:50){
  d <- sample(nrow(A), min(500, nrow(A)))
  est[i, ] <- nnls(A[d, ], b[d])$x
}
weights <- colMeans(est)
sample.weights <- weights / sum(weights)
```
```{r}
library(Hmisc)
# calculate weighted mean of read count
# this is used to calculate emission probabilities
write.table(paste('[',format(Sys.time()),'] [',sample.name,'] calculate weighted mean of read count',sep=''), file=log_file, col.names=FALSE, append=TRUE)
cat(paste('[',format(Sys.time()),']'," calculate weighted mean of read count\n",sep=''))
counts$mean <- apply(counts[, reference.samples], 
                     1, wtd.mean, sample.weights)
targets <- counts$target
write.table(paste('[',format(Sys.time()),'] [',sample.name,'] exclude probes with all zero counts',sep=''), file=log_file, col.names=FALSE, append=TRUE)
cat(paste('[',format(Sys.time()),']'," exclude probes with all zero counts\n",sep=''))
# exclude probes with all zero counts
nonzero.rows <- counts$mean > 0
nonzero.rows.df <- data.frame(target=counts$target, 
                              nonzero.rows=nonzero.rows)

counts <- counts[nonzero.rows, ]
# get the distances between consecutive probes
distances <- GetDistances(counts)
# estimate the read count variance at each probe
write.table(paste('[',format(Sys.time()),'] [',sample.name,'] estimate the read count variance at each probe',sep=''), file=log_file, col.names=FALSE, append=TRUE)
cat(paste('[',format(Sys.time()),']'," estimate the read count variance at each probe\n",sep=''))
var.estimate <- EstimateVariance(counts, reference.samples, 
                                             sample.weights)
emission.probs <- EmissionProbs(counts[, sample.name], 
                      counts$mean, var.estimate$var.estimate, 
                      counts[, "target"])
if (get.dfs){
  return(list(emission.probs=emission.probs, distances=distances))
}

head(emission.probs)
```

######################################################
# [TBD] Here need to draw the GC plots for normalized GC
######################################################
```{r}
RC_afterNorm <- counts
RC_afterNorm[1:10,c(1,2,3,4,5,6,25)]
```
```{r}
RC_beforeNorm[1:10,c(1,2,3,4,5,6,25)]
```

```{r}
RC_afterNorm_test <- RC_afterNorm[1:8000,]
dim(RC_afterNorm_test)
plot(RC_afterNorm_test$gc,RC_afterNorm_test$SP0000203, ylim = c(0,500))
```

```{r}
range(RC_afterNorm$SP0000203)
loessMod50_afterNorm <- loess(SP0000203 ~ gc, data=RC_afterNorm, span=0.50) # 50% smoothing span
# get smoothed output
smoothed50_afterNorm <- predict(loessMod50_afterNorm) 
# Plot it
ii <- ggplot(RC_afterNorm, aes(gc, smoothed50_afterNorm))
ii + geom_line()
```

```{r}
dim(RC_beforeNorm)
dim(RC_afterNorm)

range(RC_afterNorm$SP0000203)
loessMod50_afterNorm <- loess(SP0000203 ~ gc, data=RC_afterNorm, span=0.50) # 50% smoothing span
# get smoothed output
smoothed50_afterNorm <- predict(loessMod50_afterNorm) 
# Plot it
ii <- ggplot(RC_afterNorm, aes(gc, smoothed50_afterNorm))
ii + geom_line()
```

```{r}
# call CNVs with the ≈ algorithm
write.table(paste('[',format(Sys.time()),'] [',sample.name,'] call CNVs with the Viterbi algorithm',sep=''), file=log_file, col.names=FALSE, append=TRUE)
cat(paste('[',format(Sys.time()),']',' call CNVs with the Viterbi algorithm\n',sep=''))
viterbi.state <- Viterbi(emission.probs, distances, p, Tnum, D)  
viterbi.state[which(viterbi.state$viterbi.state!=2),]
# format the CNVs
write.table(paste('[',format(Sys.time()),'] [',sample.name,'] format the CNVs',sep=''), file=log_file, col.names=FALSE, append=TRUE)
cat(paste('[',format(Sys.time()),']',' format the CNVs\n',sep=''))
cnvs <- PrintCNVs(sample.name, viterbi.state, 
                       counts)
cnvs
```


